processing/src/main/java/io/druid/segment/data/GenericIndexedWriter.java

```diff
@@ -158,10 +336,129 @@ public InputStream getInput() throws IOException
+
+    int bagSizePower = bagSizePower();
+    OutputStream metaOut = Channels.newOutputStream(channel);
+    metaOut.write(GenericIndexed.VERSION_TWO);
+    metaOut.write(objectsSorted ? 0x1 : 0x0);
+    metaOut.write(Ints.toByteArray(bagSizePower));
+    metaOut.write(Ints.toByteArray(Ints.checkedCast(numWritten)));
+    metaOut.write(Ints.toByteArray(fileNameByteArray.length));
+    metaOut.write(fileNameByteArray);
+
+    try (RandomAccessFile headerFile = new RandomAccessFile(f, "r")) {
+      long previousValuePosition = 0;
+      int bagSize = 1 << bagSizePower;
+
+      int numberOfFilesRequired = (int) numWritten / bagSize;
+      if ((numWritten % bagSize) != 0) {
+        numberOfFilesRequired += 1;
+      }
+
+      try (InputStream is = new FileInputStream(ioPeon.getFile(makeFilename("values")))) {
+        int counter = -1;
+        byte[] buffer = new byte[1 << 16];
+
+        for (int i = 0; i < numberOfFilesRequired; i++) {
+          if (i != numberOfFilesRequired - 1) {
+            headerFile.seek((bagSize + counter) * Longs.BYTES); // 8 for long bytes.
+            counter = counter + bagSize;
+          } else {
+            headerFile.seek((numWritten - 1) * Longs.BYTES); // for remaining items.
+          }
+
+          long valuePosition = Long.reverseBytes(headerFile.readLong());
+          long numBytesToPutInFile = valuePosition - previousValuePosition;
+
+          try (SmooshedWriter smooshChannel = smoosher
+              .addWithSmooshedWriter(generateValueFileName(filenameBase, i), numBytesToPutInFile)) {
+            writeBytesIntoSmooshedChannel(numBytesToPutInFile, buffer, smooshChannel, is);
+            previousValuePosition = valuePosition;
+          }
+        }
+      }
+      writeHeader(smoosher, headerFile, bagSizePower);
+    }
+
+  }
+
+  public void writeToChannel(WritableByteChannel channel, FileSmoosher smoosher) throws IOException
+  {
+    if (!requireMultipleFiles) {
+      writeToChannelVersionOne(channel);
+    } else {
+      writeToChannelVersionTwo(channel, smoosher);
+    }
+  }
+
+  private void writeHeader(FileSmoosher smoosher, RandomAccessFile headerFile, int bagSizePower) throws IOException
+  {
+    ByteBuffer byteBuffer = ByteBuffer.wrap(new byte[Ints.BYTES]).order(ByteOrder.nativeOrder());
+
+    try (CountingOutputStream finalHeaderOut = new CountingOutputStream(
+        ioPeon.makeOutputStream(makeFilename("header_final")))) {
+      int numberOfElementsPerValueFile = 1 << bagSizePower;
+      long currentNumBytes = 0;
+      long relativeRefBytes = 0;
+      long relativeNumBytes;
+      headerFile.seek(0);
+
+      // following block converts long header indexes into int header indexes.
+      for (int pos = 0; pos < numWritten; pos++, --numberOfElementsPerValueFile) {
+        //conversion of header offset from long to int completed for one value file done, change relativeRefBytes
+        // to current offset.
+        if (numberOfElementsPerValueFile == 0) {
```

Comment:
IMO it's clearer to replace the logic of changing `numberOfElementsPerValueFile` with a check `pos & (numberOfElementsPerValueFile - 1) == 0`.